onnxruntime/test/shared_lib/test_inference.cc

```diff
@@ -648,6 +651,454 @@ TEST(CApiTest, multiple_varied_input_custom_op_handler) {
+      {OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC},
+      4,
+      true);
+
+  Ort::CustomOpDomain custom_op_domain("test");
+  custom_op_domain.Add(&custom_op);
+
+  Ort::SessionOptions session_options;
+  session_options.Add(custom_op_domain);
+
+  bool caught_exception = false;
+
+  try {
+    Ort::Session session(*ort_env, VARIADIC_INPUT_OUTPUT_CUSTOM_OP_MODEL_URI, session_options);
+  } catch (const Ort::Exception& excpt) {
+    caught_exception = true;
+    std::string_view exception_msg = excpt.what();
+    std::string_view expected_err = "Error Node (VariadicNode0) has output size 3 not in range [min=4, max=2147483647].";
+    ASSERT_TRUE(exception_msg.find(expected_err) != std::string_view::npos);
+  }
+
+  ASSERT_TRUE(caught_exception);
+}
+
+TEST(CApiTest, invalid_variadic_input_homogeneity_custom_op) {
+  // Create a custom op with a homogeneous variadic input. The model has heterogeneous inputs,
+  // so we expect an error.
+  TemplatedCustomOp<MyCustomEchoReversedArgsKernel> custom_op(
+      "VariadicNode",
+      // Input config
+      {ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED},
+      {OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC},
+      1,
+      true,  // Input homogeneity requirement will cause error!
+      // Output config
+      {ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED},
+      {OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC},
+      1,
+      false);
+
+  Ort::CustomOpDomain custom_op_domain("test");
+  custom_op_domain.Add(&custom_op);
+
+  Ort::SessionOptions session_options;
+  session_options.Add(custom_op_domain);
+
+  bool caught_exception = false;
+
+  try {
+    Ort::Session session(*ort_env, VARIADIC_UNDEF_INPUT_OUTPUT_CUSTOM_OP_MODEL_URI, session_options);
+  } catch (const Ort::Exception& excpt) {
+    caught_exception = true;
+    std::string_view exception_msg = excpt.what();
+    std::string_view expected_err = "Type Error: Type parameter (T0) of Optype (VariadicNode) bound to different types";
+    ASSERT_TRUE(exception_msg.find(expected_err) != std::string_view::npos);
+  }
+
+  ASSERT_TRUE(caught_exception);
+}
+
+// TODO: Remove this #if guard before merging this branch
+#if 0
+TEST(CApiTest, invalid_variadic_output_homogeneity_custom_op) {
+  // Create a custom op with a homogeneous variadic output. The model has heterogeneous outputs,
+  // so we expect an error.
+  TemplatedCustomOp<MyCustomEchoReversedArgsKernel> custom_op(
+      "VariadicNode",
+      // Input config
+      {ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED},
+      {OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC},
+      1,
+      false,
+      // Output config
+      {ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED},
+      {OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC},
+      1,
+      true);  // Output homogeneity requirement will cause error!
```

Comment:
This test does not fail as I would expect, and I'm not sure why.

We're testing for enforcement of variadic output homogeneity. If the custom op specifies a homogeneous variadic output, then I expect an error if we instantiate the custom op in a model that has outputs of different tensor element types (like below).

![variadic_output_hetero](https://user-images.githubusercontent.com/19691973/207705972-209a0eb7-63ef-4933-beb9-dbf58a93d5a6.PNG)

This same test works when testing homogeneity of the inputs (see previous test), but it doesn't work for outputs. It's as if homogeneity is not being enforced for variadic outputs. @skottmckay @hariharans29 Do you know if this is a bug or expected behavior?

Reply:
Here's some more information. I fired up the debugger and extracted the `onnx::OpSchema` calls in `CreateCustomRegistry`. For the custom op and model above, these are the schema calls.

```c++
schema.Input(
  0,         // Input index
  "Input0",  // Input name
  "",        // Description
  "T0",      // Type string
  Variadic,  // FormalParameterOption
  false,     // Homogeneous
  1          // Minimum arity
);

schema.TypeConstraint(
  "T0",                                                        // Type string
   {"tensor(float)", "tensor(double)", ..., "tensor(string)"}, // Constraints
  "all_types"                                                  // Description
);

schema.Output(
  0,         // Output index
  "Output0", // Output name
  "",        // Description
  "T0",      // Type string
  Variadic,  // FormalParameterOption
  true,      // Homogeneous (should fail validation!?)
  1          // Minimum arity
);
```

As we can see, the output is specified as homogeneous, but no validation error is thrown.

Reply:
I found the function (`Graph::InferAndVerifyTypeMatch`) in graph/graph.cc where input homogeneity is verified: https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/graph/graph.cc#L2223

It looks like the same check is missing for outputs:
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/graph/graph.cc#L2288

Reply:
I suppose the question is what should happen if the input is heterogeneous (T) and the output is homogeneous (T)? (where T is constrained to be all numeric tensor types)

Currently, the output homogeneity is not enforced, but should it? Does it even make sense to have heterogeneous input and a homogeneous output? I don't see any built-in ops with this characteristic.

Reply:
I can't think of a use-case where there would be both variadic inputs and outputs with different constraints on the types, so I would lean towards removing the test. 

For background it looks like the checking code was originally added for non-variadic inputs/outputs. if two parameters have type constraint T, and T is all tensor types, they both must use the same specific type (e.g. tensor(int64)). If all the input type constraints are correct, I think that _should_ ensure any output with the same type constraints implicitly match due to onnx type/shape inferencing propagating type info. I'm assuming that's why there's no additional check on outputs. 

When heterogeneity was added for variadic parameters I think the `if (op_formal_parameter.GetIsHomogeneous()) ` was added to skip the check.

Reply:
Thank you for the background. I removed the test.